feat: handle routing for starfyre

DeleMike opened the pull request:
from the `test_application/pages` where we define some routes we want to have, we generate the corresponding routes in the generated `dist/pages` folder.

So if we have:
``` 
test_application/
    - pages/
        - home.fyre
        - about.fyre
```
        
The corresponding generated route will be:
 
``` 
  test_application/
    - dist/
        - home.html
        - about.html
```


closes #28

DeleMike commented:
Hey @sansyrox , I hope you are well 

So far, I have been able to generate routes defined in the pages directory of the `test_application` into `dist/pages`. See [here ](https://github.com/sparckles/starfyre/pull/67#issue-1818693368) for more info

Am I on the right track? 

Also,

Few questions:
- do the generated routes need to be generated in the `test_application/dist/pages` directory?
- if we are generating routes like this, does that mean the contents in the `test_application/pages/about.fyre` needs to be copied into the file?

sansyrox commented on starfyre/router.py:
```diff
@@ -0,0 +1,58 @@
+import os
+from pathlib import Path
+
+class Router:
```
Let's call it file router as we will have an app router in the future

DeleMike replied:
Yes, thanks!

sansyrox commented on starfyre/router.py:
```diff
@@ -0,0 +1,58 @@
+    def __init__(self, pages_directory):
+        super().__init__()
+        self.pages_directory = pages_directory
+        self.generate_routes()
```
One suggestion - we don't need to call this function here as you are explicitly calling it in `__main__.py`

sansyrox reviewed (commented):
Hey @DeleMike :wave: 

Great work! :smile: :fire: 

I have left a few comments. Do let me know if you require any more information from my end :smile:

sansyrox commented:
> So far, I have been able to generate routes defined in the pages directory of the test_application into dist/pages. See https://github.com/sparckles/starfyre/pull/67#issue-1818693368for more info Am I on the right track?

Yess


Few questions:

> do the generated routes need to be generated in the test_application/dist/pages directory?

Let us create them in `test_application/dist` directory beside the `index.html` file.


> if we are generating routes like this, does that mean the contents in the test_application/pages/about.fyre needs to be copied into the file?

Not the contents, but the rendered output. i.e. the components should be resolved like they are being resolved in the `__init__.py` file.

DeleMike commented on starfyre/router.py:
```diff
@@ -0,0 +1,58 @@
+    """
+
+    def __init__(self, pages_directory):
+        super().__init__()
```
Oh, that's a mistake. Thanks for spotting it. I was in a delimma on if I need to create a Parent `Router` and rename this file as `FileBasedRouter`

DeleMike commented:
I will adjust the folder structure then. Thanks

I need more clarification on the last bit :
> Not the contents, but the rendered output. i.e. the components should be resolved like they are being resolved in the __init__.py file.

What do you mean, please?  When we are generating the new routes, the file contents needs to be parsed before it is written to this route?

sansyrox commented:
> What do you mean, please? When we are generating the new routes, the file contents needs to be parsed before it is written to this route?

e.g. `init.fyre` -> `init.py` -> `index.html`

Similary, the `about.fyre` -> `about.py` -> `about.html` 

Did this make sense? 😅 

What I am trying to say is don't copy just the contents but actually resolve the imports like we do in the init file

DeleMike commented:
Oh definitely makes sense now, thanks for clarifying! I will look into it.

DeleMike commented:
Hello @sansyrox , thanks for your feedback. I have made some progress.

I added the ability for `Starfyre` to generate the `python` files for every `fyre` file in the pages directory. The way `Starfyre` handles it is, during compilation, we check if there are some fyre  files in the `pages` directory. If there is, we ensure `Starfyre` tracks it for parsing.

so my next issue is how do I go about generating the `html` file. 

As a start, I see that generation of the `index.html` file happens when the main function of Starfyre is called.

My issue is I need the output of `return render_root(component)` from these generated python files to be sent in the generated `html` file and not the **file contents** of that python file. Since I am not triggering any main function for the routes, how do I tackle this issue of the html file generation?

As you said `test-application/pages/home.fyre` compiled to `test-application/build/home.py`  and the router generates `test-application/dist/home.html`

The `home.py` returns `render_root(component)` which will be the contents to be written into `home.html`. how do I get the result of the function?

is there another angle I should be looking at?

sansyrox commented:
Hell @DeleMike 👋 

So, if you have a look at the `__main__.py`. You will see that we are importing `app` function from the `__init__.py` in the build folder. 

Now, this file's contents are generated through the `python_transpiled_string` function.  We have to change the way that function is called.

I hope this helps. We can have a call tomorrow otherwise 😄

DeleMike commented:
Okay, sounds great! What function are you referring to here?

sansyrox commented:
> Okay, sounds great! What function are you referring to here?

@DeleMike i am talking about the 'python_transpiled_string' function

DeleMike commented:
Thanks, @sansyrox for clarifying. what do you suggest now? Do we want to stop calling that function from the `compile` function?

DeleMike commented on starfyre/file_router.py:
```diff
@@ -0,0 +1,106 @@
+    def _build_output(self, generated_routes, out_dir):
+        """
+        Transpile the output of `render(component)` to route HTML files.
+
+        This method takes a list of generated routes and an output directory and transpiles the components
+        using `render(component)`. The resulting components are then written to corresponding HTML files
+        in the output directory.
+
+        Parameters:
+            generated_routes (list): A list of route names (strings) generated from the pages directory.
+            out_dir (str): The path to the output directory where the HTML files will be written.
+
+        Example:
+            pages_directory = "test_app/pages"
+            router = FileRouter(pages_directory)
+            router.generate_routes()  # This generates the `generated_routes` list.
+            out_directory = "test_app/dist"
+            router._build_output(generated_routes, out_directory)
+        """
+        print(f'Generated routes: {generated_routes}')
+
+        root = Path(out_dir / "..").resolve()
+        app_name = (str(root).split('/'))[-1] # get the user defined project name
+
+        for route_name in generated_routes:
+            # Get the module name dynamically based on the route_name
+            module_name = f"{Path(app_name)}.build.{route_name}"
+            
+            try:
+                module = importlib.import_module(module_name)
+            except ModuleNotFoundError:
+                print(f"Error: Could not import module '{module_name}'.")
+                continue
+
+            component = module.__dict__[route_name]
+            result = str(render(component))
+
+            # write to component file
+            with open(out_dir / f"{route_name}.html", "w") as html_file:
+                html_file.write(result)
```
Hello @sansyrox , this is how I send the output of the `render` into their respective `html` files. 

What do you think of this approach?

DeleMike commented on test-application/pages/about.fyre:
Created this to test how Starfyre will behave when it encounters an empty file.

DeleMike commented on starfyre/compiler.py:
```diff
@@ -5,9 +5,14 @@
+        if entry.endswith(".fyre"):
+            fyre_files.append(entry)
+        # check inside the 'pages' folder
+        if entry == 'pages':
+            for file_ in os.listdir(project_dir / "pages"):
+                if file_.endswith(".fyre"):
+                    fyre_files.append(f'pages/{file_}')
```
Tracking all `.fyre` files defined in the `/pages` folder

sansyrox replied:
Nicee

sansyrox commented on starfyre/__main__.py:
```diff
@@ -1,5 +1,7 @@
 from starfyre import compile
 from pathlib import Path
+from .file_router import FileRouter
```
One suggestion - let us use absolute imports. i.e. `from starfyre.file_router import FileRouter`

sansyrox commented on starfyre/__main__.py:
```diff
@@ -74,12 +76,17 @@ def main(path, build):
         # At this point, the project has been compiled and the build directory has been created.
+        # Now, initialize the Router object and use it to handle file-based routing.
+        # Basically, generate the html files for the routes
+        router = FileRouter(absolute_path / "pages")
```
let us call it file_router

sansyrox commented on starfyre/exceptions.py:
```diff
@@ -16,4 +16,5 @@ class UnknownTagError(Exception):
-   pass
\ No newline at end of file
+   def __init__(self, message='Unknown tag found.'):
+    super().__init__(message)
```
Can you please explain why we need this?

DeleMike replied:
Okay, thanks fo asking!

This line calls the `__init__` method of the parent class, `Exception`. By doing so, the custom exception class inherits the behavior of the built-in `Exception` class and ensures that the exception is properly initialized with the provided error message.

I believe it makes the code a bit easier to understand. We know that it is dependent on the `Exception` class.

sansyrox commented on starfyre/file_router.py:
```diff
@@ -0,0 +1,106 @@
+from starfyre import create_component, render
```
Need to go through this file. Will do by tomorrow morning 😅

sansyrox reviewed (commented):
Hey @DeleMike 👋 

Thank you for the PR! I have some suggestions.

Do let me know if you require any more information from my end 😄

sansyrox commented on starfyre/file_router.py:
```diff
@@ -0,0 +1,106 @@
+            router.generate_routes()  # This generates the routes and corresponding HTML files.
+
+        Raises:
+            FileNotFoundError: If the specified pages directory does not exist.
```
Love the docstring here! 🔥

sansyrox commented on starfyre/file_router.py:
```diff
@@ -0,0 +1,106 @@
+            except ModuleNotFoundError:
+                print(f"Error: Could not import module '{module_name}'.")
+                continue
+
+            component = module.__dict__[route_name]
+            result = str(render(component))
+
+            # write to component file
+            with open(out_dir / f"{route_name}.html", "w") as html_file:
+                html_file.write(result)
```
We would have to change some things here - 

We should generate the python file using this function - https://github.com/sparckles/starfyre/blob/9b2f863863d8a7c007ab79490e0e1945639b96eb/starfyre/compiler.py#L159

And all the logic that you have created above should be called in the compile function - https://github.com/sparckles/starfyre/blob/9b2f863863d8a7c007ab79490e0e1945639b96eb/starfyre/compiler.py#L190


Finally, all the python files in the `pages` folder should be treated like entry points and we will have to the `render_root` function and not the `render` function. As they are roots of a new file.

We will have to alter the logic in this file - https://github.com/sparckles/starfyre/blob/9b2f863863d8a7c007ab79490e0e1945639b96eb/starfyre/compiler.py#L100

DeleMike replied:
> We would have to change some things here -
> 
> We should generate the python file using this function -
> 
> https://github.com/sparckles/starfyre/blob/9b2f863863d8a7c007ab79490e0e1945639b96eb/starfyre/compiler.py#L159

I am generating the python files from this function. In order to generate the python files, all I did was to adjust the `get_fyre_files` function.

DeleMike replied:
> And all the logic that you have created above should be called in the compile function -
> 
> https://github.com/sparckles/starfyre/blob/9b2f863863d8a7c007ab79490e0e1945639b96eb/starfyre/compiler.py#L190
> 
> Finally, all the python files in the `pages` folder should be treated like entry points and we will have to the `render_root` function and not the `render` function. As they are roots of a new file.
> 
> We will have to alter the logic in this file -
> 
> https://github.com/sparckles/starfyre/blob/9b2f863863d8a7c007ab79490e0e1945639b96eb/starfyre/compiler.py#L100

Thanks for the great feedback, @sansyrox 🔥
Just some questions to understand more:

Can you explain what you mean "all the logic I have created above"? Are you saying I should move the `_build_output` logic into `compile`?

<hr>

Also, can you explain what you mean in the last section, I believe it is `fyre` files that ought to be in the `pages` directory. And where would I stop to user `render`? is it here:

```python
component = module.__dict__[route_name]
result = str(render(component))
# write to component file
with open(out_dir / f"{route_name}.html", "w") as html_file:
    html_file.write(result)
```

sansyrox replied:
@DeleMike , I think it would be better to have a meeting at 7pm, Monday. Does that work for you? I will explain all of this on a call.

DeleMike replied:
That's great @sansyrox. However, can we move it to 8pm GMT+1? I have Stand Ups by 7pm every Mondays.

sansyrox replied:
Sounds good.

DeleMike replied:
Settled then 👍🏾. Thanks

sansyrox commented:
Hey @DeleMike 👋 

I have left a few comments. Let me know if you require any more information from me 😄

DeleMike commented on starfyre/__main__.py:
```diff
@@ -13,42 +15,27 @@ def write_js_file(path):
-        f.write("<script src='store.js'></script>")
-        f.write(app)
-"""
-        )
+"""if __name__ == '__main__':
+    pass
+    """)
```
I removed the route generation from this function. This function will be used to call `write_to_js` function and also used to create `build/__main.py__` file which will serve as the entry point for the application.

DeleMike commented on starfyre/compiler.py:
I added new exception classes. 
1) `InitFyreMissingError` - used to check if the root `__init__.fyre` file is missing. We will not compile app

2) `IndexFileConflictError` - used to check if the user defined a route called `index.fyre`. Such a file is not allowed, as it would conflict with the generation of the main 'index.html' file that is produced from the transpilation of the `__init__.fyre` file.

DeleMike commented on starfyre/compiler.py:
```diff
@@ -1,13 +1,23 @@
+            for file_ in os.listdir(project_dir / "pages"):
+                if file_.endswith(".fyre"):
+                    # check for the invalid 'index.fyre'
+                    if file_.lower() == 'index.fyre':
+                        raise IndexFileConflictError()
```
Check if the user created an `index.fyre` file in the pages directory

sansyrox replied:
Niceee

DeleMike commented:
Hello @sansyrox , thanks for your feedback always! I have been able to use them to refactor the code:

Here are 3 key highlights:
- made the `generate_routes` function to be responsible solely for route generations. From generating `index.html` to all routes defined in the pages directory
- refactored `create_main_file` to have a job of calling the `write_to_js` function and also removed all code for generating `index` route
- prevented user from creating `index.fyre` in the pages folder and ensured that starfyre compilation **fails** if `__init__.fyre `does not exist"

Finally, can this PR be broken into two? I want us to break into `FileBasedRouting` then another PR to focus on `AppBasedRouting`.

DeleMike commented on starfyre/file_router.py:
Here what I did here in order to ensure that all route generation is done in one function is:

add an additional route "app" to the generated routes generated_routes.insert(0, 'app')
Then performed the imports dynamically based on whether it is the root 'app' component or not.
finally write to file.

sansyrox commented:
Hey @DeleMike 👋 

I will have a proper look tomorrow morning 😄 

But 

> Finally, can this PR be broken into two? I want us to break into FileBasedRouting then another PR to focus on AppBasedRouting.

Yes, please!

sansyrox commented on starfyre/__main__.py:
```diff
@@ -13,42 +15,27 @@ def write_js_file(path):
-        f.write(app)
-"""
-        )
+"""if __name__ == '__main__':
+    pass
+    """)
```
One tip - this line doesn't do anything

DeleMike replied:
Are you suggesting I create an empty file for this? The `build/__main__.py` file is important. I noticed while testing locally.

sansyrox replied:
Yes.

sansyrox commented on starfyre/compiler.py:
```diff
@@ -202,7 +212,11 @@ def compile(entry_file_name):
+    fyre_files = get_fyre_files(project_dir)
+
+    # check if __init__.fyre exist else stop compilation
+    if '__init__.fyre' not in fyre_files:
+        raise InitFyreMissingError()
```
nicee

DeleMike commented:
> Hey @DeleMike 👋
> 
> I will have a proper look tomorrow morning 😄
> 
> But
> 
> > Finally, can this PR be broken into two? I want us to break into FileBasedRouting then another PR to focus on AppBasedRouting.
> 
> Yes, please!

Alright then. That means when we are through with file based routing this PR can be merged, yeah?

sansyrox commented on starfyre/file_router.py:
```diff
@@ -0,0 +1,116 @@
+            router._build_output(generated_routes, out_directory)
+        """
+        # index.html is to be generated first since it is the entry point of the app
+        generated_routes.insert(0, 'app')
```
we should not mutate a param that is being passed. Let us create a new copy of this array.

sansyrox commented on starfyre/file_router.py:
```diff
@@ -0,0 +1,116 @@
+            else:
+                component_key = route_name
+            
+            module_name = f"{Path(app_name)}.build.{route_name}" if not route_name == 'app' else f"{Path(app_name)}.build"
```
this line is doing a lot of things. Can we please split it into multiple parts?

sansyrox commented on starfyre/file_router.py:
```diff
@@ -0,0 +1,116 @@
+            try:
+                module = importlib.import_module(module_name)
+                component = module.__dict__[component_key]
+                result = str(render_root(component)) if not route_name == 'app' else component
```
should we not be using `render_root` for every page?

sansyrox commented on starfyre/file_router.py:
```diff
@@ -0,0 +1,116 @@
+        # read the contents from the generated python files
+        self._build_output(generated_routes=routes, out_dir=dist_dir)
+
+    def _build_output(self, generated_routes, out_dir):
```
This function should not be present in a Router. A Router should not be responsible for building the files. That is the purpose of the compiler area.

This part should be present in the `__main__` file, which is being populated using the `create_main_file` function.

The `__main__` file is responsible for converting the IR(intermediate representation, i.e. python in our case) to the final output, which is JS and HTML in our case.

I would recommend moving it to the `create_main_file` function and renaming the function to something else.

DeleMike commented on starfyre/__main__.py:
Hello @sansyrox, thanks for your feedback - they make so much sense. Can you look into this?

I have created a function called `build_routes_output` to be responsible for converting the IR(Python file) to the final output(HTML, JS).

When reviewing, please focus on the `build/__main__.py`. Am I on the right track?

Everything works well but check out the file. This is a cut from the result after I run the build function:

```python
if __name__ == '__main__':
    _build_output(generated_routes=['about', 'home'], path="/XXXX/XXXX/XXXX/XXXX/test-application")
```

So the above says we have '`about`', and '`home`' routes available and `path` points to where our Starfyre project is.

sansyrox commented on starfyre/__main__.py:
```diff
@@ -14,43 +16,69 @@ def write_js_file(path):
 from pathlib import Path
+import importlib
+
+def _build_output(generated_routes, path):
```
Two suggestions - 
- Can we use a different name here? Something like `generate_pages` or `create_pages` or on these lines?

- Can you please add a docstring?

sansyrox replied:
And I suppose this logic will be changed after our discussion today.

sansyrox commented on starfyre/file_router.py:
```diff
@@ -0,0 +1,63 @@
+import importlib
+
+
+class FileRouter:
+    """
+    A router that handles file-based routing.
+
+    This router parses the specified pages directory to automatically generate routes based on
+    the file names. Each file in the pages directory is treated as a separate route. 
+
+    Parameters:
+        pages_directory (str): The path to the directory containing the pages files.
+    Example:
+        pages_directory = "test-application/pages"
+        file_router = FileRouter(pages_directory)
+    """
```
This docstring should be a part of the `__init__` function, no?

sansyrox replied:
Not resolved?

DeleMike replied:
Oh, I'm not sure if I understand this question well.

I added docstrings to the missing `__init__` function.

The docstring below
```python
class FileRouter:
```
is for the class `FileRouter`. 

Do you mean I should take all the docstring below that class inside the `__iniy__` function?

sansyrox replied:
Yes

sansyrox commented on starfyre/file_router.py:
```diff
@@ -0,0 +1,63 @@
+
+    def generate_routes(self):
+        """
+        Generate routes and create corresponding HTML files.
```
we don't create html files anymore here.

We can possibly name it as `populate_router`?

sansyrox reviewed (commented):
Hey @DeleMike 👋 

Thank you for your PR 😄 

I have some suggestions. Do let me know if you have nay questions.

DeleMike commented:
Thanks, Sanskar. I will adjust accordingly.


On Tue, 8 Aug 2023, 23:13 Sanskar Jethi, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
>
> Hey @DeleMike <https://github.com/DeleMike> 👋
>
> Thank you for your PR 😄
>
> I have some suggestions. Do let me know if you have nay questions.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sparckles/starfyre/pull/67#pullrequestreview-1568342594>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AL2L2ETEWDZCSKWINVEW443XUK2YHANCNFSM6AAAAAA2VZCHXU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

sansyrox commented on starfyre/__main__.py:
```diff
@@ -14,42 +16,78 @@ def write_js_file(path):
+        try:
+            module = importlib.import_module(module_name)
+            print(f"Module is = {{module}}")
+            component = module.__dict__[component_key]
+            result = str(render_root(component)) if not route_name == 'app' else component
```
Wait, why call render root again? I thought this was already being generated at every page?

Also is there an alternative of using `__dict__`?

DeleMike commented:
Hello @sansyrox, 

I have worked on the fie router more. 
Currently we agreed to remove this
```python
# result = str(render_root(component)) if not route_name == 'app' else component
result = component # since, during transpilation, component will already be processed as 'render_root(component)'
```

So after,  I removed it and adjusted the function `python_transpiled_string` to have only this as its output, irrespective of whether we are currently in the `__init__.fyre` file or any other file.
```python
return f'''
from starfyre import create_component, render_root

def fx_{root_name}():
    # not nesting the code to preserve the frames
    component = create_component("""
{pyml_lines}
""", css="""
{css_lines}
""", js="""
{js_lines}
""", client_side_python="""
{client_side_python}
""",
component_name="""{root_name}"""
)
    return render_root(component)

{root_name}=fx_{root_name}()
'''
```
so because the parser cannot identify the output of the above as components, the parser throws `UnknownTagError` for all our custom components
```
starfyre.exceptions.UnknownTagError: Unknown tag: "store". Please review line 1 in your "app" component in the pyml code.
```

You can see that it doesn't recognise `store` as a custom component anymore and I suspect the same for other custom components like `parent`.

I have tried a scenario if in the parser we can simply get the local_variables in the `__init__.fyre` file, like this
```python
if self.component_name == 'app':
  print(f'Type of local is {type(self.local_variables)}')
  for key, value in self.local_variables.items():
      print(f'Key = {key}')
      print(f'Value = {value}')
      self.custom_components[key] = value # I tried adding the value of the local variables into this new dictionary.
```
Because of the idea above, we will add another check in the `handle_starttag` and `handle_endtag` to check the `custom_components` if the particular tag exists like: 
The below is only for `handle_starttag`

```python
self.custom_components = {}
if tag not in self.generic_tags and tag in self.custom_components:
    component = self.custom_components[tag]
    print('Fetched value for tag={tag} = component={component}')
    tag = component.tag # here is where an issue is, I suspect even the the others will have the same issue since component is now a string.
    component.props = {**component.props, **props}
    component.state = {**component.state, **state}
    component.event_listeners = {**component.event_listeners, **event_listeners}
    self.stack.append(component)

    return
```

and here is the error message after this change:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/runpy.py", line 187, in _run_module_as_main
    mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
  File "/usr/local/lib/python3.10/runpy.py", line 146, in _get_module_details
    return _get_module_details(pkg_main_name, error)
  File "/usr/local/lib/python3.10/runpy.py", line 110, in _get_module_details
    __import__(pkg_name)
  File "/home/akin/software_eng/starfyre/test-application/build/__init__.py", line 49, in <module>
    app=fx_app()
  File "/home/akin/software_eng/starfyre/test-application/build/__init__.py", line 24, in fx_app
    component = create_component("""
  File "/home/akin/software_eng/starfyre/starfyre/__init__.py", line 21, in create_component
    parser.feed(pyml)
  File "/usr/local/lib/python3.10/html/parser.py", line 110, in feed
    self.goahead(0)
  File "/usr/local/lib/python3.10/html/parser.py", line 170, in goahead
    k = self.parse_starttag(i)
  File "/usr/local/lib/python3.10/html/parser.py", line 344, in parse_starttag
    self.handle_starttag(tag, attrs)
  File "/home/akin/software_eng/starfyre/starfyre/parser.py", line 137, in handle_starttag
    tag = component.tag
AttributeError: 'str' object has no attribute 'tag'
```

What do we do from here? Do we stick to the former process? or is this something that will cause an issue in the near or far future?


BTW, I have adjusted the code base to use this for this section of the `generate_pages` in the main function to:
```diff
- component = module.__dict__[component_key]
+ component = getattr(module, component_key, f"{{component_key}} does not exist")
```

It remains this issue with `render_root` and the parser. What am I missing?

sansyrox commented:
@DeleMike 👋 

> I have worked on the fie router more.
Currently we agreed to remove this# result = str(render_root(component)) if not route_name == 'app' else componentresult = component # since, during transpilation, component will already be processed as 'render_root(component)'So after, I removed it and adjusted the function python_transpiled_string to have only this as its output, irrespective of whether we are currently in the __init__.fyre file or any other file.

We should be keeping the imports like we are currently doing in the `__init__.py` file. This error is occurring as we don't have the imports.

DeleMike commented:
> We should be keeping the imports like we are currently doing in the `__init__.py` file. This error is occurring as we don't have the imports.

Thanks for the feedback. Please, I need more clarification. What file are you referring to that I keep the imports like the `__init__.py` file?

Can you explain more about how the absence of imports affects this issue?

sansyrox commented:
> Thanks for the feedback. Please, I need more clarification. What file are you referring to that I keep the imports like the __init__.py file?

Yes.

>  Unknown tag: "store". Please review line 1 in your "app" component in the pyml code.

The file is unable to resolve the `store` variable according to this error.

DeleMike commented:
Hello @sansyrox, thanks for the clarification earlier.

I have another update. I have been able to tackle the issue where the parser could not resolve the custom components. e.g `<store>, <parent>`. The issue was not totally the parser's fault.
```python
from .parent import parent
from .store import store
```
I used the Python debugger to peek deeply into the codebase and I saw that our parser was able to resolve the imports. 

The issue was with how we compiled the code. The parser would read the contents of each imported file but it could not resolve if it was a `Component`. It always classified all our imported components as non-Components and that was the issue.

To resolve this issue and ensure that `extract_components` was actually able to resolve our custom-defined components children of the `Component` class.

I resorted to this in the python_transpiled_string function:
```python
def python_transpiled_string(
    pyml_lines, css_lines, js_lines, client_side_python, file_name
):
    file_name = file_name.replace(".py", "").split("/")[-1]
    pyml_lines = "".join(pyml_lines)
    css_lines = "".join(css_lines)
    js_lines = "".join(js_lines)
    client_side_python = "".join(client_side_python)

    root_name = None

    if "__init__" in file_name:
        root_name = "app"
    else:
        root_name = file_name

    if root_name == "app":
        return f'''
from starfyre import create_component, render_root

def fx_{root_name}():
    # not nesting the code to preserve the frames
    component = create_component("""
{pyml_lines}
""", css="""
{css_lines}
""", js="""
{js_lines}
""", client_side_python="""
{client_side_python}
""",
component_name="""{root_name}"""
)
    return render_root(component)

{root_name}=fx_{root_name}()
'''
    else:
        return f'''
from starfyre import create_component, render_root

def fx_{root_name}():
    component = create_component("""
{pyml_lines}
""", css="""
{css_lines}
""", js="""
{js_lines}
""", client_side_python="""
{client_side_python}
""",
component_name="""{root_name}"""
)
    return component

{root_name}=fx_{root_name}()
rendered_{root_name} = render_root({root_name}) # see the change here
'''
```

So now, we ensure every custom-defined is classified as a `Component`.
And to add to it, instead of handling the rendering in the `generate_pages` function, we now get the output from other components with `rendered_{root_name} = render_root({root_name})`. The result of `rendered_{root_name}` will have the HTML output for the component. 

With this, we make each page a root component and ensure that the parser can resolve the components properly.

sansyrox commented on starfyre/compiler.py:
```diff
@@ -153,6 +163,7 @@ def fx_{root_name}():
     return component
```
why not just do this
```
return render_root(component)

rendered_{root_name} = fx_{root_name}()
```

Also, I don't think that I understand it completely. I will have a proper read tomorrow morning 😄

DeleMike replied:
It is okay to do it this way:
```python
return component
```
but we will have to resort to the former way of rendering the pages. Like this:
```python
result = str(render_root(component)) if not route_name == 'app' else component # we don't want this.
``` 


So now what I have done is:
- ensure that each file parsed is treated as a component(i.e a `Component` child)
- ensure we have an output of `render_root` for each of our pages that are being transpiled.

Why?
It will help the parser be able to resolve & recognise our imports. I noticed that the parser could actually read the contents of the file while I was debugging but it could not resolve these imports(`store` and `parent`) as children of the `Component` class, which they are. This was what was causing the issue.

So that's why I decided that, we let each page being transpiled to be a normal `Component` but also keep the output of `render_root(page_being_transpiled)`. This output will be used to write to the individual pages.

sansyrox replied:
>  I noticed that the parser could actually read the contents of the file while I was debugging but it could not resolve these imports(store and parent) as children of the Component class, which they are

What was the issue here?

sansyrox commented on starfyre/__main__.py:
```diff
@@ -9,34 +9,87 @@
-    And the `__init__.py` file will have a component that will render the root component. It will be named `app`.
 
-    You can have a look at the `test-application/build/__init__.py` file to see what it looks like.
+def build_routes_output(generated_routes, path):
```
Hey @DeleMike 👋 

Can you please add a docstring here?

DeleMike replied:
Done.

sansyrox commented on starfyre/__main__.py:
```diff
@@ -57,26 +110,29 @@ def main(path, build):
-        create_main_file(str(absolute_path))
         # Start/run project
         subprocess.run(
             [sys.executable, "-m", "build"],
             cwd=path,
-            stdout=subprocess.PIPE,
+            # stdout=subprocess.PIPE,
             stderr=None,
         )
 
```
I have some changes on the main for this. Can you fetch the latest changes from the main branch here?

sansyrox commented on starfyre/__main__.py:
```diff
@@ -9,34 +9,87 @@
-"""if __name__ == '__main__':
-    pass
-    """)
+            f'''from starfyre import create_component, render_root
```
I wonder if we can have this is a python file somewhere. This string is getting a bit hard to review 😅

DeleMike replied:
Okay, I have broken it into another file

sansyrox replied:
Amazing, thank you 😄

sansyrox commented on starfyre/__main__.py:
```diff
@@ -9,34 +9,87 @@
+    for route_name in generated_routes:
+        print(f'route name is = {{route_name}}')
+        
+        if route_name.lower() == 'app':
+            component_key = 'app'  # For the 'app' component in `build/pages/__init__.py`
+        else:
+            component_key = route_name
+        
+        if route_name == 'app':
+            module_name = f"{{Path(app_name)}}.build.pages"
+        else:
+            module_name= f"{{Path(app_name)}}.build.pages.{{route_name}}"
+        
+        try:
+            module = importlib.import_module(module_name)
+            if component_key == 'app':
+                component = getattr(module, component_key, f"{{component_key}} does not exist")
+            else:
+                component = getattr(module, f'rendered_{{component_key}}')
+            result = str(component)
+        except ModuleNotFoundError:
+            print(f"Error: Could not import module '{{module_name}}'.")
+            continue
```
Can you please explain this to me?

DeleMike replied:
The code block above is responsible for dynamically importing the correct component based on the route name and generating the HTML content for each route using that component.

```python
if route_name.lower() == 'app':
    component_key = 'app'  # For the 'app' component in `build/pages/__init__.py`
else:
    component_key = route_name
```

First, we check if the route name is 'app', and the code checks if the route_name is equal to 'app'. If it is, it sets component_key to 'app'. This is because, in the `__init__.py` file inside the build/pages directory, there might be a component named 'app'. This is a special case.


We then get the module name. We now define all compiled Python files(IR files) into the pages directory.
after this, we try to get the output of the function which is the rendered output which happened through the compiler.

and if the module is not found we skip that step though we might want to throw an error saying that we should properly import our files.

DeleMike commented:
Hey @sansyrox, thanks for your feedback! In this latest update, I have answered the questions you asked and I have reduced the `starfyre/__main__.py` file contents by introducing a template file `starfyre/main_template.py` file. This will be used to enhance code readability.

sansyrox commented on starfyre/__main__.py:
```diff
@@ -1,55 +1,51 @@
-    The main file is also responsible for adding the `store.js` file to the `index.html` file.
+    This function gets the HTML output for each generated route based on the provided `generated_routes` list.
+    It also creates the main execution file (__main__.py) that orchestrates rendering and writing HTML content to files.
+    The rendered HTML content is generated by importing and utilizing the corresponding components for each route.
```
```suggestion
    The rendered HTML content is generated by importing and utilizing the corresponding components for each route in the __main__ file.
```

sansyrox commented on starfyre/__main__.py:
```diff
@@ -1,55 +1,51 @@
-    directory = Path(path_ ) / ".." / "dist"
-    if not directory.exists():
-        directory.mkdir()
+    output_file_path = path + "/build/__main__.py"
```
```suggestion
    main_file_path = path + "/build/__main__.py"
```

sansyrox commented on starfyre/__main__.py:
```diff
@@ -1,55 +1,51 @@
-"""
-        )
+    # create empty __init__.py file
+    with open(init_file_path, "w"):
```
We can use `Path.touch` here instead of this 😅

sansyrox commented on starfyre/__main__.py:
```diff
@@ -1,55 +1,51 @@
+        pass
 
+    with open(output_file_path, "w") as f:
+        f.write(main_script_content)
```
```suggestion
        f.write(main_file_content)
```

sansyrox commented on starfyre/__main__.py:
```diff
@@ -70,15 +66,23 @@ def main(path, build):
 
     # Convert path to absolute path
     absolute_path = Path(path).resolve()
+    print(f'Absoulte path = {absolute_path}')
```
```suggestion
    print(f'Absolute path of the project = {absolute_path}')
```

sansyrox commented:
Hey @DeleMike 👋 

The `__init__.fyre` should be in the `pages` folder. That would also change some logic in the import resolution 😅

sansyrox commented on starfyre/main_template.py:
Can we give this file a more specific name please? `main_template` is a bit too generic.

sansyrox replied:
@DeleMike this too

DeleMike replied:
Done

sansyrox commented on starfyre/main_template.py:
```diff
@@ -0,0 +1,76 @@
+def get_main_template(user_routes, path):
```
same for this function

sansyrox replied:
@DeleMike 👋 

These comments need to be addressed too 😄

DeleMike replied:
Done

sansyrox commented on starfyre/__main__.py:
```diff
@@ -1,55 +1,51 @@
 
-
-def create_main_file(path):
+def build_routes_output(generated_routes, path):
```
can we name this function as `prepare_html_and_main()`

sansyrox commented:
Hey @DeleMike 👋 

Thank you for your updates. I have some suggestions. Please do let me know if you have any questions 😄

DeleMike commented:
Hello @sansyrox , thanks for the feedback, I have applied them. The most notable will be ensuring that `__init__fyre` exists in the `pages` directory amongst other changes. Thanks!

sansyrox reviewed (approved):
LGTM , @DeleMike 🔥 

Great work. I just want to do a final local test and will merge it afterwards

DeleMike commented:
Thanks for the feedback @sansyrox. Fingers crossed! 🤞🏾

sansyrox commented:
All LGTM. Great work @DeleMike 🔥
